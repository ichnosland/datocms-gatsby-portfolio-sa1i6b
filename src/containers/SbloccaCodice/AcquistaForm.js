/**
*
* AcquistaForm
*
*/
import React from 'react';
import PropTypes from 'prop-types';
import { Field, reduxForm } from 'redux-form/immutable';
import { Button as ButtonPaypal } from 'paypal-checkout';

import ReactDOM from 'react-dom';

import AlertBanner from 'components/AlertBanner';
import Div from 'components/Div';
import { Button } from 'components/Button';
import { H3 } from 'components/Heading';
import { BrandTxt } from 'components/Text';
import { Separator, SeparatorText } from 'components/Paragraph';
import { CardForm, FormElement } from 'components/FormElements';
import ButtonRadioCheck from 'components/FormElements/ButtonRadioCheck';
import { PAYPAL } from 'configuration';
import validate from './validator';

export const required = (value) => value ? undefined : 'Campo obbligatorio';
const PayPalButtonComponent = ButtonPaypal.driver('react', { React, ReactDOM });

export class AcquistaForm extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  render() {
    const {
      feedback,
      togglePaypalBox,
      prodotti,
      selectedValue,
      authorizePaypal,
      paymentPaypal,
      userAnagraphics,
    } = this.props;

    const prodottoSelezionato = prodotti.filter(
      (prodotto) => (prodotto.order_data === selectedValue)
    )[0] || undefined;

    return (
      <CardForm>
        {feedback.hasFeedback && feedback.messaggio ? (
          <AlertBanner
            text={feedback.messaggio}
            actioncolor={feedback.tipologia}
          >
            {feedback.messaggio}
          </AlertBanner>
        ) :
          ([
            !prodottoSelezionato && (
              <div key="scelta_prodotto">
                <H3 margin="1em 0">Seleziona i prodotti che desideri acquistare:</H3>
                <FormElement>
                  {prodotti.map((prodotto) => (
                    <Field
                      htmlFor="prodotto"
                      name="prodotto"
                      id={`prodotto_${prodotto.id}`}
                      key={`prodotto_${prodotto.id}`}
                      component={ButtonRadioCheck}
                      validate={[required]}
                      type="radio"
                      value={prodotto.order_data}
                      label={`${prodotto.descrizione} ${prodotto.prezzo}€`}
                    />
                  ))}
                </FormElement>
              </div>
            ),
            prodottoSelezionato && (
              <div key="prodotto_selezionato">
                <Div bgColor="white" radius="3px" padding="10px">
                  <p><strong><BrandTxt>Prodotto selezionato:</BrandTxt></strong><br />{prodottoSelezionato.descrizione}<br /><br /><strong><BrandTxt>Prezzo:</BrandTxt></strong><br />{prodottoSelezionato.prezzo}€</p>
                </Div>
                <Separator>
                  <SeparatorText>Procedi all&apos;acquisto</SeparatorText>
                </Separator>
                <FormElement>
                  <PayPalButtonComponent
                    commit
                    style={PAYPAL.style}
                    locale={PAYPAL.locale}
                    env={PAYPAL.env}
                    validate={this.paypalValidate}
                    client={PAYPAL.client}
                    payment={/* istanbul ignore next */ (data, actions) => paymentPaypal(
                      data, actions, {
                        total: prodottoSelezionato.prezzo,
                        currency: PAYPAL.currency,
                        description: `${prodottoSelezionato.descrizione} - utente: ${userAnagraphics.email}`,
                      },
                    )}
                    onAuthorize={
                      /* istanbul ignore next */ (data, actions) => authorizePaypal(data, actions, prodottoSelezionato.order_data)
                    }
                  />
                </FormElement>
                <Separator>
                  <SeparatorText>Oppure</SeparatorText>
                </Separator>
              </div>
            ),
            <Button full outline={1} onClick={togglePaypalBox} type="button" key="button_annulla">Annulla e inserisci codice</Button>,
          ])
        }
      </CardForm>
    );
  }
}

AcquistaForm.propTypes = {
  feedback: PropTypes.shape({
    hasFeedback: PropTypes.bool.isRequired,
    tipologia: PropTypes.string.isRequired,
    messaggio: PropTypes.string.isRequired,
  }).isRequired,
  togglePaypalBox: PropTypes.func.isRequired,
  prodotti: PropTypes.arrayOf(
    PropTypes.shape({
      prezzo: PropTypes.string.isRequired,
      descrizione: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired,
      order_data: PropTypes.string.isRequired,
    })
  ).isRequired,
  selectedValue: PropTypes.string,
  authorizePaypal: PropTypes.func.isRequired,
  paymentPaypal: PropTypes.func.isRequired,
  userAnagraphics: PropTypes.shape({
    email: PropTypes.string.isRequired,
  }).isRequired,
};

export default reduxForm({
  form: 'acquista',
  validate,
})(AcquistaForm);
