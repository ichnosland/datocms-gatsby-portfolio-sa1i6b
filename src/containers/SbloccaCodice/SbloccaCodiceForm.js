/**
*
* SbloccaCodiceForm
*
*/
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Field, reduxForm } from 'redux-form/immutable';
import { createTextMask } from 'redux-form-input-masks';

import AlertBanner from 'components/AlertBanner';
import { Button } from 'components/Button';
import Img from 'components/Image';
import { H3 } from 'components/Heading';
import P, { Separator, SeparatorText, LegalInfo } from 'components/Paragraph';
import { CardForm, FormElement } from 'components/FormElements';
import renderedField from 'components/FormElements/Field';
import validate from './validator';
import paybox from './images/payboxinfo.png';


export const FormTitle = styled(H3)`
  margin: 1em 0;
`;

FormTitle.defaultProps = {
  theme: {
    brand: '#00BBEF',
  },
};

export const maskDefinitions = {
  A: {
    regExp: /[A-Za-z0-9]/,
    transform: (char) => (char.toUpperCase()),
  },
};

const codeTextMask = createTextMask({
  pattern: 'AAAA-AAAA-AAAA',
  maskDefinitions,
});

export const required = (value) => value ? undefined : 'Campo obbligatorio';


export class SbloccaCodiceForm extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  render() {
    const { feedback, handleSubmit, valid, togglePaypalBox } = this.props;

    return (
      <CardForm onSubmit={handleSubmit}>
        <FormTitle>Se possiedi già un codice</FormTitle>
        <P>Inserisci qui il codice ricevuto dal tuo docente o quello che trovi sul frontespizio del libro.</P>
        <P>Assicurati di inserire tutti e 12 i caratteri da cui è composto.</P>
        {feedback.hasFeedback && (
          <AlertBanner
            text={feedback.messaggio}
            actioncolor={feedback.tipologia}
          >
            {feedback.messaggio}
          </AlertBanner>
        )}
        <FormElement>
          <Field
            name="codice"
            component={renderedField}
            type="text"
            label="codice"
            validate={[required]}
            {...codeTextMask}
            align="center"
          />
        </FormElement>
        <FormElement>
          <Button
            actioncolor="okay"
            full
            disabled={!valid}
          >Attiva Premium
          </Button>
        </FormElement>
        <Separator>
          <SeparatorText>Oppure</SeparatorText>
        </Separator>
        <FormTitle>Procedi all&apos;acquisto</FormTitle>
        <P>Assicurati di acquistare una licenza Premium solo se sei stato invitato a far parte di una classe dal tuo docente.</P>
        <FormElement>
          <Button outline={1} full onClick={togglePaypalBox} type="button">Acquista Premium</Button>
        </FormElement>
        <Img src={paybox} alt="PayPal" responsive />
        <LegalInfo>PayPal è il metodo rapido e sicuro per pagare e farsi pagare online</LegalInfo>
      </CardForm>
    );
  }
}

SbloccaCodiceForm.propTypes = {
  feedback: PropTypes.shape({
    hasFeedback: PropTypes.bool.isRequired,
    tipologia: PropTypes.string.isRequired,
    messaggio: PropTypes.string.isRequired,
  }).isRequired,
  handleSubmit: PropTypes.func.isRequired,
  valid: PropTypes.bool.isRequired,
  togglePaypalBox: PropTypes.func.isRequired,
};

export default reduxForm({
  form: 'sbloccaCodice',
  validate,
})(SbloccaCodiceForm);
