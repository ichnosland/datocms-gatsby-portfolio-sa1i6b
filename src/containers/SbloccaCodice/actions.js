/*
 *
 * SbloccaCodice actions
 *
 */

import {
  SBLOCCACODICE_SBLOCCA_POST,
  SBLOCCACODICE_SPINNER_SET,
  SBLOCCACODICE_FEEDBACK_SET,
  SBLOCCACODICE_FEEDBACK_RESET,
  SBLOCCACODICE_ACQUISTA_SET,
  SBLOCCACODICE_ACQUISTA_RESET,
  SBLOCCACODICE_ACQUISTA_FETCH,
  SBLOCCACODICE_ACQUISTA_POST,
} from './constants';

export const sbloccaCodiceSbloccaPost = (codice, configuration, history) => ({
  type: SBLOCCACODICE_SBLOCCA_POST,
  codice,
  configuration,
  history,
});

export const sbloccaCodiceSpinnerSet = (enable) => ({
  type: SBLOCCACODICE_SPINNER_SET,
  enable,
});

export const sbloccaCodiceFeedbackSet = (enable, tipologia, messaggio) => ({
  type: SBLOCCACODICE_FEEDBACK_SET,
  enable,
  tipologia,
  messaggio,
});

export const sbloccaCodiceFeedbackReset = () => ({
  type: SBLOCCACODICE_FEEDBACK_RESET,
});

export const sbloccaCodiceAcquistaSet = (payload) => ({
  type: SBLOCCACODICE_ACQUISTA_SET,
  payload,
});

export const sbloccaCodiceAcquistaReset = () => ({
  type: SBLOCCACODICE_ACQUISTA_RESET,
});

export const sbloccaCodiceAcquistaFetch = (disciplina) => ({
  type: SBLOCCACODICE_ACQUISTA_FETCH,
  disciplina,
});

export const sbloccaCodiceAcquistaPost = (paypalId, orderData, configuration, ticketData) => ({
  type: SBLOCCACODICE_ACQUISTA_POST,
  paypalId,
  orderData,
  configuration,
  ticketData,
});
