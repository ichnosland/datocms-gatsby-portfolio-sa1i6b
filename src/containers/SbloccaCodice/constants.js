/*
 *
 * SbloccaCodice constants
 *
 */

export const SBLOCCACODICE_URI_SBLOCCA = 'cicero_academy/api/codicisblocco';
export const SBLOCCACODICE_URI_ACQUISTA = 'cicero_academy/api/codicisblocco/acquista';

export const SBLOCCACODICE_SBLOCCA_POST = 'app/containers/SbloccaCodice/SBLOCCACODICE_SBLOCCA_POST';
export const SBLOCCACODICE_SPINNER_SET = 'app/containers/SbloccaCodice/SBLOCCACODICE_SPINNER_SET';
export const SBLOCCACODICE_FEEDBACK_SET = 'app/containers/SbloccaCodice/SBLOCCACODICE_FEEDBACK_SET';
export const SBLOCCACODICE_FEEDBACK_RESET = 'app/containers/SbloccaCodice/SBLOCCACODICE_FEEDBACK_RESET';
export const SBLOCCACODICE_ACQUISTA_FETCH = 'app/containers/SbloccaCodice/SBLOCCACODICE_ACQUISTA_FETCH';
export const SBLOCCACODICE_ACQUISTA_POST = 'app/containers/SbloccaCodice/SBLOCCACODICE_ACQUISTA_POST';
export const SBLOCCACODICE_ACQUISTA_SET = 'app/containers/SbloccaCodice/SBLOCCACODICE_ACQUISTA_SET';
export const SBLOCCACODICE_ACQUISTA_RESET = 'app/containers/SbloccaCodice/SBLOCCACODICE_ACQUISTA_RESET';
