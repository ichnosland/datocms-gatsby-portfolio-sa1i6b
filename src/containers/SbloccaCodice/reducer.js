/*
 *
 * SbloccaCodice reducer
 *
 */

import { combineReducers } from 'redux-immutable';
import { fromJS } from 'immutable';

import {
  SBLOCCACODICE_SPINNER_SET,
  SBLOCCACODICE_FEEDBACK_SET,
  SBLOCCACODICE_FEEDBACK_RESET,
  SBLOCCACODICE_ACQUISTA_SET,
  SBLOCCACODICE_ACQUISTA_RESET,
} from './constants';

export const defaultSbloccaCodiceFeedback = fromJS({
  hasFeedback: false,
  tipologia: '',
  messaggio: '',
});

export const defaultSbloccaCodicePaypal = fromJS({
  prodottiDisponibili: [],
  display: false,
  isLoaded: false,
});

function sbloccaCodiceSpinnerReducer(state = fromJS(false), action) {
  switch (action.type) {
    default:
      return state;

    case SBLOCCACODICE_FEEDBACK_SET:
      return fromJS(false);

    case SBLOCCACODICE_SPINNER_SET:
      return fromJS(action.enable);
  }
}

function sbloccaCodiceFeedbackReducer(state = defaultSbloccaCodiceFeedback, action) {
  switch (action.type) {
    default:
      return state;

    case SBLOCCACODICE_FEEDBACK_SET:
      return state.merge({
        hasFeedback: action.enable,
        tipologia: action.tipologia,
        messaggio: action.messaggio,
      });

    case SBLOCCACODICE_FEEDBACK_RESET:
      return defaultSbloccaCodiceFeedback;
  }
}

function sbloccaCodicePaypalReducer(state = defaultSbloccaCodicePaypal, action) {
  switch (action.type) {
    default:
      return state;

    case SBLOCCACODICE_ACQUISTA_SET:
      return state.merge(action.payload);

    case SBLOCCACODICE_ACQUISTA_RESET:
      return defaultSbloccaCodicePaypal;
  }
}

const sbloccaCodiceReducer = combineReducers({
  spinner: sbloccaCodiceSpinnerReducer,
  feedback: sbloccaCodiceFeedbackReducer,
  paypal: sbloccaCodicePaypalReducer,
});

export default sbloccaCodiceReducer;
