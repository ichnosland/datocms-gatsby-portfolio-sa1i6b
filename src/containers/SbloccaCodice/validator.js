export default (values) => {
  const errors = {};

  if (!values.get('codice')) {
    errors.codice = 'Campo obbligatorio';
  }

  return errors;
};
