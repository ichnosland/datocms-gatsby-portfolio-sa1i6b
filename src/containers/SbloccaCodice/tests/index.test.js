import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'configureStore';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { reset } from 'redux-form/immutable';
import { fromJS } from 'immutable';

import injectReducer from 'utils/injectReducer';

import Spinner from 'components/Spinner';
import TopBar from 'components/TopBar';
import AlertBanner from 'components/AlertBanner';
import reducerCurryer from 'common/applications/reducer';
import { modalSetData } from 'containers/ModalBox/actions';
import userReducer from 'containers/User/reducer';
import SbloccaCodice, {
  SbloccaCodiceView,
} from '../index';
import sbloccaCodiceReducer, { defaultSbloccaCodiceFeedback } from '../reducer';
import {
  sbloccaCodiceSbloccaPost,
  sbloccaCodiceAcquistaSet,
  sbloccaCodiceAcquistaPost,
  sbloccaCodiceAcquistaFetch,
  sbloccaCodiceFeedbackReset,
} from '../actions';
import SbloccaCodiceForm from '../SbloccaCodiceForm';
import AcquistaForm from '../AcquistaForm';
import validate from '../validator';


const configuration = {
  product: 'product',
  hasPremium: true,
  disciplinaId: 123,
  payPalButtonId: '123456789',
};
const store = configureStore({}, {});
const history = {
  push: () => { },
};

const sbloccaProps = {
  configuration,
  sbloccaCodice: {
    spinner: false,
    feedback: {
      hasFeedback: false,
      tipologia: 'tipologia',
      messaggio: 'messaggio',
    },
    paypal: {
      prodottiDisponibili: [{
        id: 1,
        prezzo: 3.4,
        descrizione: 'descrizione',
        order_data: 'order_data_1',
      }, {
        id: 2,
        prezzo: 5.5,
        descrizione: 'scrizionede',
        order_data: 'order_data_2',
      }],
      display: true,
    },
  },
  formSblocco: {
    values: {
      codice: '123-codice',
    },
  },
  history,
  formAcquista: {},
  onResetFormSbloccaCodice: () => { },
  onResetFormAcquista: () => { },
  onSbloccaCodiceAcquistaFetch: () => { },
  onSbloccaCodiceAcquistaSet: () => { },
  onSbloccaCodiceAcquistaPost: () => { },
  onSbloccaCodiceFeedbackReset: () => { },
  onModalSetData: () => { },
  onModalSetEmptyData: () => { },
  onSendCodice: () => { },
  userAnagraphics: {
    first_name: 'nome',
    last_name: 'cognome',
    email: 'acme@acme.com',
  },
  userAppData: {
    docente: false,
  },
};


describe('<SbloccaCodiceView />', () => {
  it('visualizza un container con uno spinner al suo interno e paypal.display == false', () => {
    const props = {
      ...sbloccaProps,
      sbloccaCodice: {
        ...sbloccaProps.sbloccaCodice,
        spinner: true,
        paypal: {
          ...sbloccaProps.sbloccaCodice.paypal,
          display: false,
        },
      },
    };
    const renderedComponent = shallow(
      <SbloccaCodiceView {...props} />
    );
    expect(renderedComponent.find(TopBar).length).toBe(1);
    expect(renderedComponent.find(SbloccaCodiceForm).length).toBe(0);
    expect(renderedComponent.find(AcquistaForm).length).toBe(0);
    expect(renderedComponent.find(Spinner).length).toBe(1);
    expect(renderedComponent.find(AlertBanner).length).toBe(0);
    expect(renderedComponent).toMatchSnapshot();
  });

  it('visualizza un container con uno spinner e paypal.display == true mostra messaggio', () => {
    const props = {
      ...sbloccaProps,
      sbloccaCodice: {
        ...sbloccaProps.sbloccaCodice,
        spinner: true,
        paypal: {
          ...sbloccaProps.sbloccaCodice.paypal,
          display: true,
        },
      },
    };
    const renderedComponent = shallow(
      <SbloccaCodiceView {...props} />
    );
    expect(renderedComponent.find(TopBar).length).toBe(1);
    expect(renderedComponent.find(SbloccaCodiceForm).length).toBe(0);
    expect(renderedComponent.find(AcquistaForm).length).toBe(0);
    expect(renderedComponent.find(Spinner).length).toBe(1);
    expect(renderedComponent.find(AlertBanner).length).toBe(1);
    expect(renderedComponent).toMatchSnapshot();
  });

  it('visualizza il SbloccaCodiceForm', () => {
    const props = {
      ...sbloccaProps,
      sbloccaCodice: {
        ...sbloccaProps.sbloccaCodice,
        paypal: {
          ...sbloccaProps.sbloccaCodice.paypal,
          display: false,
        },
      },
    };
    const renderedComponent = shallow(
      <SbloccaCodiceView {...props} />
    );
    expect(renderedComponent.find(TopBar).length).toBe(1);
    expect(renderedComponent.find(SbloccaCodiceForm).length).toBe(1);
    expect(renderedComponent.find(AcquistaForm).length).toBe(0);
    expect(renderedComponent.find(Spinner).length).toBe(0);
    expect(renderedComponent.find(AlertBanner).length).toBe(0);
  });

  it('visualizza AcquistaForm', () => {
    const props = {
      ...sbloccaProps,
      sbloccaCodice: {
        ...sbloccaProps.sbloccaCodice,
        paypal: {
          ...sbloccaProps.sbloccaCodice.paypal,
          display: true,
        },
      },
    };
    const renderedComponent = shallow(
      <SbloccaCodiceView {...props} />
    );
    expect(renderedComponent.find(TopBar).length).toBe(1);
    expect(renderedComponent.find(SbloccaCodiceForm).length).toBe(0);
    expect(renderedComponent.find(AcquistaForm).length).toBe(1);
    expect(renderedComponent.find(Spinner).length).toBe(0);
    expect(renderedComponent.find(AlertBanner).length).toBe(0);
  });

  it('SbloccaCodiceForm.handleSubmit deve richiamare la prop onSendCodice', () => {
    const props = {
      ...sbloccaProps,
      sbloccaCodice: {
        ...sbloccaProps.sbloccaCodice,
        paypal: {
          ...sbloccaProps.sbloccaCodice.paypal,
          display: false,
        },
      },
      onSendCodice: jest.fn(),
      onResetFormSbloccaCodice: jest.fn(),
    };
    const renderedComponent = shallow(
      <SbloccaCodiceView {...props} />
    );

    expect(props.onSendCodice).not.toHaveBeenCalled();
    renderedComponent.find(SbloccaCodiceForm).props().handleSubmit({ preventDefault: () => { } });
    expect(props.onSendCodice).toHaveBeenCalledWith(
      props.formSblocco.values.codice,
      configuration,
      props.history
    );
    expect(props.onResetFormSbloccaCodice).toHaveBeenCalled();
  });

  it('SbloccaCodiceForm.togglePaypalBox se i dati non sono stati caricati', () => {
    const props = {
      ...sbloccaProps,
      sbloccaCodice: {
        ...sbloccaProps.sbloccaCodice,
        paypal: {
          ...sbloccaProps.sbloccaCodice.paypal,
          display: false,
        },
      },
      onSbloccaCodiceAcquistaFetch: jest.fn(),
      onSbloccaCodiceAcquistaSet: jest.fn(),
      onSbloccaCodiceFeedbackReset: jest.fn(),
    };
    const renderedComponent = shallow(
      <SbloccaCodiceView {...props} />
    );

    const instance = renderedComponent.instance();
    expect(props.onSbloccaCodiceAcquistaFetch).not.toHaveBeenCalled();
    expect(props.onSbloccaCodiceFeedbackReset).not.toHaveBeenCalled();
    expect(props.onSbloccaCodiceAcquistaFetch).not.toHaveBeenCalled();
    instance.togglePaypalBox();
    expect(props.onSbloccaCodiceAcquistaFetch).toHaveBeenCalledWith(123);
    expect(props.onSbloccaCodiceFeedbackReset).not.toHaveBeenCalled();
    expect(props.onSbloccaCodiceAcquistaSet).toHaveBeenCalledWith({
      display: true,
    });
  });

  it('SbloccaCodiceForm.togglePaypalBox se i dati sono giÃ  caricati', () => {
    const props = {
      ...sbloccaProps,
      sbloccaCodice: {
        ...sbloccaProps.sbloccaCodice,
        paypal: {
          ...sbloccaProps.sbloccaCodice.paypal,
          display: false,
          isLoaded: true,
        },
      },
      onSbloccaCodiceAcquistaFetch: jest.fn(),
      onSbloccaCodiceAcquistaSet: jest.fn(),
      onSbloccaCodiceFeedbackReset: jest.fn(),
    };
    const renderedComponent = shallow(
      <SbloccaCodiceView {...props} />
    );

    const instance = renderedComponent.instance();
    expect(props.onSbloccaCodiceAcquistaFetch).not.toHaveBeenCalled();
    expect(props.onSbloccaCodiceFeedbackReset).not.toHaveBeenCalled();
    expect(props.onSbloccaCodiceAcquistaFetch).not.toHaveBeenCalled();
    instance.togglePaypalBox();
    expect(props.onSbloccaCodiceAcquistaFetch).not.toHaveBeenCalled();
    expect(props.onSbloccaCodiceFeedbackReset).toHaveBeenCalled();
    expect(props.onSbloccaCodiceAcquistaSet).toHaveBeenCalledWith({
      display: true,
    });
  });

  it('SbloccaCodiceForm.togglePaypalBox se devo nascondere acquista', () => {
    const props = {
      ...sbloccaProps,
      sbloccaCodice: {
        ...sbloccaProps.sbloccaCodice,
        paypal: {
          ...sbloccaProps.sbloccaCodice.paypal,
          display: true,
          isLoaded: true,
        },
      },
      onSbloccaCodiceAcquistaFetch: jest.fn(),
      onSbloccaCodiceAcquistaSet: jest.fn(),
      onSbloccaCodiceFeedbackReset: jest.fn(),
    };
    const renderedComponent = shallow(
      <SbloccaCodiceView {...props} />
    );

    const instance = renderedComponent.instance();
    expect(props.onSbloccaCodiceAcquistaFetch).not.toHaveBeenCalled();
    expect(props.onSbloccaCodiceFeedbackReset).not.toHaveBeenCalled();
    expect(props.onSbloccaCodiceAcquistaFetch).not.toHaveBeenCalled();
    instance.togglePaypalBox();
    expect(props.onSbloccaCodiceAcquistaFetch).not.toHaveBeenCalled();
    expect(props.onSbloccaCodiceFeedbackReset).toHaveBeenCalled();
    expect(props.onSbloccaCodiceAcquistaSet).toHaveBeenCalledWith({
      display: false,
    });
  });

  it('paymentPaypal deve restituire la crezione delle transaction', () => {
    const renderedComponent = shallow(
      <SbloccaCodiceView {...sbloccaProps} />
    );

    const actions = {
      payment: {
        create: jest.fn(),
      },
    };
    const instance = renderedComponent.instance();
    instance.paymentPaypal(undefined, actions, {
      total: 123,
      currency: 'EUR',
      description: 'description',
    });
    expect(actions.payment.create).toHaveBeenCalledWith({
      transactions: [{
        amount: {
          total: 123,
          currency: 'EUR',
        },
        description: 'description',
      }],
    });
  });

  it('authorizePaypal deve restituire i dati del pagamento', () => {
    const props = {
      ...sbloccaProps,
      onSbloccaCodiceAcquistaPost: jest.fn(),
      onResetFormAcquista: jest.fn(),
    };
    const renderedComponent = shallow(
      <SbloccaCodiceView {...props} />
    );

    const mockThen = (data) => (data({ id: 'PAYPAL_ID' }));
    const actions = {
      payment: {
        execute: () => ({ then: mockThen }),
      },
    };
    const instance = renderedComponent.instance();
    instance.authorizePaypal(undefined, actions, 'order_data');
    expect(props.onSbloccaCodiceAcquistaPost).toHaveBeenCalledWith(
      'PAYPAL_ID', 'order_data', configuration, {
        email: 'acme@acme.com',
        isDocente: false,
        nome: 'nome cognome',
      }
    );
    expect(props.onResetFormAcquista).toHaveBeenCalled();
  });
});

describe('<SbloccaCodice />', () => {
  const withReducerConfiguration = injectReducer({ key: 'configuration', reducer: reducerCurryer(configuration) });
  const withReducerSbloccaCodice = injectReducer({ key: 'sbloccaCodice', reducer: sbloccaCodiceReducer });
  const withUserReducer = injectReducer({ key: 'user', reducer: userReducer });

  const ConfigurationWrapper = compose(
    withReducerConfiguration,
    withReducerSbloccaCodice,
    withUserReducer,
    connect()
  )(SbloccaCodice);

  it('controllo che le props iniziali del componente siano settate correttamente', () => {
    const renderedComponent = shallow(
      <ConfigurationWrapper history={history} />, { context: { store } }
    ).dive().dive().dive().dive().dive().dive().dive();
    const receivedProps = renderedComponent.props();

    expect(receivedProps.sbloccaCodice.feedback).toEqual(defaultSbloccaCodiceFeedback.toJS());
    expect(receivedProps.sbloccaCodice.spinner).toEqual(false);
  });

  it('controllo che onSendCodice faccia il dispatch di sbloccaCodiceSbloccaPost', () => {
    const mockStore = store;
    mockStore.dispatch = jest.fn();
    const renderedComponent = shallow(
      <ConfigurationWrapper history={history} />, { context: { store: mockStore } }
    ).dive().dive().dive().dive().dive().dive().dive();
    const receivedProps = renderedComponent.props();

    expect(mockStore.dispatch).not.toHaveBeenCalled();
    receivedProps.onSendCodice(sbloccaProps.formSblocco.values.codice, configuration);
    expect(mockStore.dispatch).toHaveBeenCalledWith(
      sbloccaCodiceSbloccaPost(sbloccaProps.formSblocco.values.codice, configuration)
    );
  });

  it('controllo mapDispatchToProps', () => {
    const mockStore = store;
    mockStore.dispatch = jest.fn();
    const renderedComponent = shallow(
      <ConfigurationWrapper history={history} />, { context: { store: mockStore } }
    ).dive().dive().dive().dive().dive().dive().dive();
    const receivedProps = renderedComponent.props();

    receivedProps.onSendCodice(1, 2, 3);
    expect(mockStore.dispatch).toHaveBeenCalledWith(
      sbloccaCodiceSbloccaPost(1, 2, 3)
    );

    receivedProps.onResetFormSbloccaCodice();
    expect(mockStore.dispatch).toHaveBeenCalledWith(
      reset('sbloccaCodice')
    );

    receivedProps.onResetFormAcquista();
    expect(mockStore.dispatch).toHaveBeenCalledWith(
      reset('acquista')
    );

    receivedProps.onSbloccaCodiceFeedbackReset();
    expect(mockStore.dispatch).toHaveBeenCalledWith(
      sbloccaCodiceFeedbackReset()
    );

    receivedProps.onSbloccaCodiceAcquistaFetch(1);
    expect(mockStore.dispatch).toHaveBeenCalledWith(
      sbloccaCodiceAcquistaFetch(1)
    );

    receivedProps.onSbloccaCodiceAcquistaSet({ data: 123 });
    expect(mockStore.dispatch).toHaveBeenCalledWith(
      sbloccaCodiceAcquistaSet({ data: 123 })
    );

    receivedProps.onSbloccaCodiceAcquistaPost(1, 'orderData', 4);
    expect(mockStore.dispatch).toHaveBeenCalledWith(
      sbloccaCodiceAcquistaPost(1, 'orderData', 4)
    );

    receivedProps.onModalSetData({ data: 123 });
    expect(mockStore.dispatch).toHaveBeenCalledWith(
      modalSetData({ data: 123 })
    );
  });
});


describe('<SbloccaCodiceView />', () => {
  it('controllo che il comportamento del validate sia corretto', () => {
    expect(validate(fromJS({ codice: '1234' }))).toEqual({});
    expect(validate(fromJS({ codice: undefined }))).toEqual({
      codice: 'Campo obbligatorio',
    });
  });
});
