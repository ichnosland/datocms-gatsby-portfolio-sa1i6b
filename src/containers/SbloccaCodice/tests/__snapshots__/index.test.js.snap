// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SbloccaCodiceView /> visualizza un container con uno spinner al suo interno e paypal.display == false 1`] = `
<Page>
  <TopBar
    backArrow={
      Object {
        "enabled": false,
        "url": "",
      }
    }
    backNav={
      Object {
        "enabled": true,
        "url": undefined,
      }
    }
    closeBtn={
      Object {
        "enabled": false,
        "url": "",
      }
    }
    lesson={
      Object {
        "enabled": false,
        "url": "",
      }
    }
    pinned={true}
    single={true}
    title="Premium"
    topSwitcher={
      Object {
        "items": Array [],
      }
    }
  />
  <Container>
    <Div
      padding="0 0 20px;"
    >
      <Spinner
        key="spinner"
        number={3}
      />
    </Div>
    <ButtonGroup
      buttons={
        Array [
          Object {
            "id": 1,
            "label": "Assistenza",
            "onClickFunction": [Function],
          },
        ]
      }
    />
  </Container>
</Page>
`;

exports[`<SbloccaCodiceView /> visualizza un container con uno spinner e paypal.display == true mostra messaggio 1`] = `
<Page>
  <TopBar
    backArrow={
      Object {
        "enabled": false,
        "url": "",
      }
    }
    backNav={
      Object {
        "enabled": true,
        "url": undefined,
      }
    }
    closeBtn={
      Object {
        "enabled": false,
        "url": "",
      }
    }
    lesson={
      Object {
        "enabled": false,
        "url": "",
      }
    }
    pinned={true}
    single={true}
    title="Premium"
    topSwitcher={
      Object {
        "items": Array [],
      }
    }
  />
  <Container>
    <Div
      padding="0 0 20px;"
    >
      <Spinner
        key="spinner"
        number={3}
      />
      <AlertBanner
        actioncolor="action"
        direction="column"
        icona="hourglass"
        key="alertBanner"
        margin="20px auto"
        maxWidth="380px"
        padding="30px 0 0"
        theme={
          Object {
            "brand": "#00BBEF",
            "radius": Object {
              "general": "6px",
            },
          }
        }
      >
        <Button__Icon
          height="18px"
          path={
            <path
              d="M15.448 0H3.511c-.57 0-1.085.316-1.345.824-.26.507-.212 1.11.122 1.572L7.07 9l-4.782 6.604a1.504 1.504 0 0 0-.122 1.572c.26.508.775.824 1.345.824h11.937c.57 0 1.086-.316 1.345-.824.26-.507.213-1.11-.122-1.572L11.89 9l4.781-6.604a1.504 1.504 0 0 0 .122-1.572A1.504 1.504 0 0 0 15.448 0zm0 .692a.818.818 0 0 1 .663 1.298L11.035 9l5.076 7.01a.818.818 0 0 1-.663 1.298H3.511a.818.818 0 0 1-.662-1.298L7.925 9 2.849 1.99A.818.818 0 0 1 3.51.692h11.937zm-.773 14.864a.633.633 0 0 0-.1-.205l-3.297-2.406c-.886-.704-2.71-.704-3.596 0L4.384 15.35a.633.633 0 0 0-.119.416.631.631 0 0 0 .628.582h9.174a.64.64 0 0 0 .223-.04.626.626 0 0 0 .405-.542.613.613 0 0 0-.02-.211zM6.493 5.604L8.446 8.3c.51.704 1.558.704 2.068 0l1.953-2.697H6.493z"
              fillRule="nonzero"
            />
          }
          theme={
            Object {
              "brand": "#00BBEF",
            }
          }
          viewBox="0 0 18 18"
          width="18px"
        />
        <p>
          Stiamo processando il pagamento:
          <br />
          non chiudere il browser e attendi
          <br />
          di essere reindirizzato. Grazie
        </p>
      </AlertBanner>
    </Div>
    <ButtonGroup
      buttons={
        Array [
          Object {
            "id": 1,
            "label": "Assistenza",
            "onClickFunction": [Function],
          },
        ]
      }
    />
  </Container>
</Page>
`;
